src/html/HTMLWriter.jl

```diff
@@ -745,12 +746,67 @@ const _SEARCHRECORD_IGNORED_BLOCK_TYPES = Union{
     Documenter.DocsNodesBlock,
     Documenter.SetupNode,
 }
-function searchrecord(ctx::HTMLContext, navnode::Documenter.NavNode, node::Node)
-    # Skip indexing special at-blocks
-    if node.element isa _SEARCHRECORD_IGNORED_BLOCK_TYPES
-        return nothing
+
+# Content segment for grouping content by sections
+struct ContentSegment
+    section_header::Union{Node, Nothing}  # AnchoredHeader node or nothing for default section
+    content_nodes::Vector{Node}           # Content nodes belonging to this section
+end
+
+# Segment page content by sections for improved search indexing
+function segment_page_by_sections(page_mdast::Node)
+    segments = ContentSegment[]
+    current = ContentSegment(nothing, Node[])
+
+    for node in page_mdast.children
+        if node.element isa Documenter.AnchoredHeader
+            # Save previous segment if it has content or a header
+            if !isempty(current.content_nodes) || current.section_header !== nothing
+                push!(segments, current)
+            end
+            # Start new section
+            current = ContentSegment(node, Node[])
+        else
+            # Skip nodes that shouldn't be indexed
+            if !(node.element isa _SEARCHRECORD_IGNORED_BLOCK_TYPES)
+                push!(current.content_nodes, node)
+            end
+        end
+    end
+
+    # Add final segment
+    if !isempty(current.content_nodes) || current.section_header !== nothing
+        push!(segments, current)
+    end
+
+    return segments
+end
+
+# Create search record for a content segment
+function searchrecord(ctx::HTMLContext, navnode::Documenter.NavNode, segment::ContentSegment)
+    if segment.section_header === nothing
+        # Default section - use page title and aggregate content
+        content_text = join([mdflatten(node) for node in segment.content_nodes], "\n\n")
+        return SearchRecord(
+            ctx, navnode;
+            fragment = "",  # no fragment for default section
+            category = "section",
+            text = content_text
+        )
+    else
+        # Named section
+        a = segment.section_header.element.anchor
+        section_title = mdflatten(segment.section_header)
+        content_text = isempty(segment.content_nodes) ? "" :
+            join([mdflatten(node) for node in segment.content_nodes], "\n\n")
+        return SearchRecord(
+            ctx, navnode;
+            fragment = Documenter.anchor_fragment(a),
+            title = section_title,
+            category = "section",
+            text = content_text
+        )
     end
```

Comment:
You could also merge the two `SearchRecord(...)` calls now resp. after my other proposed changes. Something like
```suggestion
        fragment = Documenter.anchor_fragment(a)
    end
    return SearchRecord(
        ctx, navnode;
        fragment,
        title = section_title,
        category = "section",
        text = content_text
    )
```
and then replace the `SearchRecord` call before the `else` by something like
```
        section_title = nothing
        fragment = ""   # no fragment for default section
```

(Actually I'd also rename the `section_title` variable to `title`)